docs/gss.md

```diff
@@ -0,0 +1,305 @@
+
+```json
+{
+    "properties": {
+        "name": {
+            "type": "string",
+            "maxLength": 100,
+            "required": true,
+            "hint": "my.package.messages.nameHint",
+            "errors": {
+                "": "my.package.messages.nameInvalid",
+                "maxLength": "my.package.messages.nameTooLong"
+            }
+        },
+        "email": {
+            "type": "string",
+            "required": true,
+            "format": "email",
+            "hint": "my.package.messages.emailHint",
+            "errors": {
+                "" : "my.package.messages.emailInvalid"
+            }
+        },
+        "badgeColor": {
+            "enum": ["#ff0000", "#00ff00", "#0000ff"],
+            "enumLabels": ["my.package.messages.badgeRed", "my.package.messages.badgeGreen", "my.package.messages.badgeBlue"]
+        }
+    }
+}
+```
+
+Let's talk first about the `hint` element.  A single `hint` is allowed for a given element, and the value is expected to
+be an i18n message key.  In an HTML UI, this might be displayed near the field to be entered using an html `<label>`
+element.
+
+The `errors` element is somewhat more complex, as an element may have failed validation for a number of reasons.  The
+`errors` element is an object, each of whose keys directly matches a rule for the described element.  So, in the above
+`name` example, we provide a specific message for cases in which the `maxLength` rule is broken.
+
+There is also a mechanism to indicate a message key to use for any validation failure for which we don't have more
+specific information.  In the above examples, failing to include a `name` element will result in a general validation
+error message being displayed.  If no `errors` information is provided for a given element, this package provides
+default messages for each type of rule which are used instead of the raw output from the underlying validator (see
+below).
+
+The `enumLabels` element is specifically designed to allow internationalising the text displayed for things like radio
+groups or drop-down lists.  Using the above message bundle and GSS Schema, we might generate a drop-down list to input a
+`badgeColor` that looks something like:
+
+```html
+<select name="badgeColor">
+    <option value="#ff0000">Red</option>
+    <option value="#00ff00">Green</option>
+    <option value="#0000ff">Blue</option>
+</select>
+```
+
+## Validator
+
+Although we use a particular industry standard JSON Schema validator ([AJV](https://github.com/epoberezkin/ajv)), the
+validation functions provided in this package standardise validation errors to conform to our own independent format, as
+shown here:
+
+```json
+
+{
+    "isValid": false,
+    "errors": [
+        {
+            "dataPath": ["requiredField"],
+            "schemaPath": ["properties", "requiredField", "required"],
+            "rule": { "required": true },
+            "message": "validation-required-field-missing"
```

Comment:
This style of message key could be adjusted so it is consistent with the kinds of message keys shown in the previous section describing the "errors" schema field

Reply:
I (lazily) let this package use non-namespaced keys for its own messages, so that is the actual full key at the moment.  Seems like I should just namespace that with gpii.schema.messages?

Reply:
That does sound reasonable - although my primary original question was about consistency in the docs itself - the earlier message bundle uses package qualification plus camelCasing e.g. my.package.messages.emailHint - whereas this section using hyphens and no packages. As well as the actual namespacing fix, we should pick a single style for examples and impl

Reply:
However, we should revisit again our earlier discussions with the Bulgarians about I18N formatting and our "various forms of LOADERS" referenced in the following pad: https://pad.gpii.net/p/i18n-meetings-6294n3p - For example the mere fact that you have some stuff in a globally namespaced constant should count for something, and then again we also want to make life easier for users of particularly frequently-referenced messages.

Reply:
Yes, we may need to chat through this one again, as my current approach (passing in a bundle with namespaced keys) is not really compatible with using a namespaced key that can be resolved to a template by way of a global variable.  I've committed a take on this that at least standardises the namespacing and casing for further discussion.